missing rule: enforce multiple var declarations [$30]

Hi. I've tried to review each and every rule of JSHint, JSCS and ESLint to see if I could entirely "get rid of" the first two in favour of the latter one. As suggested by the owner of the @geteslint twitter account, I'm reporting the JSCS rules I've found actually not covered by ESLint. My apologies if it's a duplicate.

 - disallowMultipleVarDecl

I know there's the `one-var` rule, but what about enforcing the contrary? I.e enforcing a `var` for each declaration.

Actually JSCS has three settings :
 - `true` disallows multiple variable declarations except within a for loop
 - 'strict' disallows all multiple variable declarations
 - 'exceptUndefined' allows declarations where all variables are not defined, e.g


 ```js
 var a, b; // valid with 'exceptUndefined'
 var x = 1;  // valid with 'true', 'strict' and 'exceptUndefined'
 var y = 2;  // valid with 'true', 'strict' and 'exceptUndefined'
 for (var i = 0, j = arr.length; i < j; i++) {}  // valid with `true' and `exceptUndefined'

 var x = 1, // invalid
     y = 2; // invalid
 var x, y = 2, z; // invalid
 ```



<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/7465930-missing-rule-enforce-multiple-var-declarations?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

@xjamundx I wrote a test for
```js
function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }
```
but it fails on `=`. I'm assuming it doesn't support destructuring?
```
  1 failing

  1) one-var function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }:
     AssertionError: Should have no errors but had 1: [ { fatal: true,
    severity: 2,
    message: 'Unexpected token =',
    line: 1,
    column: 51 } ]: expected 1 to equal 0
```

you have to add `{ecmaFeatures: { destructuring: true }` into the config
for that rule!

On Wed, Mar 18, 2015 at 9:36 PM, Danny Fritz <notifications@github.com>
wrote:

> @xjamundx <https://github.com/xjamundx> I wrote a test for
>
> function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }
>
> but it fails on =. I'm assuming it doesn't support destructuring?
>
>   1 failing
>
>   1) one-var function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }:
>      AssertionError: Should have no errors but had 1: [ { fatal: true,
>     severity: 2,
>     message: 'Unexpected token =',
>     line: 1,
>     column: 51 } ]: expected 1 to equal 0
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eslint/eslint/issues/1619#issuecomment-83318577>.
>

--- a/docs/rules/one-var.md
+++ b/docs/rules/one-var.md
@@ -13,6 +13,8 @@ function foo() {
 
 This rule is aimed at preventing a possible misunderstanding about scoping of variables and to enforce a single variable declaration convention. As such, it will warn when it encounters more than one variable declaration statement in a function scope.
 
+When configured with `"always"` as the first option (the default), the following patterns are considered warnings:
+
 The following patterns are considered warnings:
 
 ```js
@@ -44,6 +46,41 @@ function foo() {
 
     if (baz) {
         qux = true;
+    }
+}
+```
+
+When configured with `"never"` as the first option, the following patterns are considered warnings:
+
+```js
+function foo() {
+    var bar,
+        baz;
+}
+
+function foo() {
+    var bar,
+        qux;
+
+    if (baz) {
+        qux = true;
+    }
+}
+```
+
+The following patterns are not considered warnings:
+
+```js
+function foo() {
+    var bar;
+    var baz;
+}
+
+function foo() {
+    var bar;
+
+    if (baz) {
+        var qux = true;
     }
 }
 ```

--- a/lib/rules/one-var.js
+++ b/lib/rules/one-var.js
@@ -1,6 +1,7 @@
 /**
  * @fileoverview A rule to ensure the use of a single variable declaration.
  * @author Ian Christian Myers
+ * @copyright 2015 Danny Fritz. All rights reserved.
  * @copyright 2013 Ian Christian Myers. All rights reserved.
  */
 
@@ -12,25 +13,43 @@
 
 module.exports = function(context) {
 
+    var MODE = context.options[0] || "always";
+
     //--------------------------------------------------------------------------
     // Helpers
     //--------------------------------------------------------------------------
 
     var functionStack = [];
 
+    /**
+     * Increments the functionStack counter.
+     * @returns {void}
+     * @private
+     */
     function startFunction() {
         functionStack.push(false);
     }
 
+    /**
+     * Decrements the functionStack counter.
+     * @returns {void}
+     * @private
+     */
     function endFunction() {
         functionStack.pop();
     }
 
-    function checkDeclarations(node) {
+    /**
+     * Determines if there is more than one var statement in the current scope.
+     * @returns {boolean} Returns true if it is the first var declaration, false if not.
+     * @private
+     */
+    function hasOnlyOneVar() {
         if (functionStack[functionStack.length - 1]) {
-            context.report(node, "Combine this with the previous 'var' statement.");
+            return true;
         } else {
             functionStack[functionStack.length - 1] = true;
+            return false;
         }
     }
 
@@ -44,7 +63,18 @@ module.exports = function(context) {
         "FunctionExpression": startFunction,
         "ArrowFunctionExpression": startFunction,
 
-        "VariableDeclaration": checkDeclarations,
+        "VariableDeclaration": function(node) {
+            var declarationCount = node.declarations.length;
+            if (MODE === "never") {
+                if (declarationCount > 1) {
+                    context.report(node, "Split 'var' declaration into multiple statements.");
+                }
+            } else {
+                if (hasOnlyOneVar()) {
+                    context.report(node, "Combine this with the previous 'var' statement.");
+                }
+            }
+        },
 
         "Program:exit": endFunction,
         "FunctionDeclaration:exit": endFunction,

--- a/tests/lib/rules/one-var.js
+++ b/tests/lib/rules/one-var.js
@@ -19,7 +19,26 @@ eslintTester.addRuleTest("lib/rules/one-var", {
     valid: [
         "function foo() { var bar = true; }",
         "function foo() { var bar = true, baz = 1; if (qux) { bar = false; } }",
-        "var foo = function () { var bar = true; baz(); }"
+        "var foo = function () { var bar = true; baz(); }",
+        {
+            code: "function foo() { var bar = true, baz = false; }",
+            args: [2, "always"]
+        },
+        {
+            code: "function foo() { var bar = true; var baz = false; }",
+            args: [2, "never"]
+        },
+        {
+            code: "function foo() { var bar = true; var baz = false; }",
+            args: [2, "never"]
+        },
+        {
+            code: "function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }",
+            ecmaFeatures: {
+                destructuring: true
+            },
+            args: [2, "never"]
+        }
     ],
     invalid: [
         {
@@ -80,6 +99,33 @@ eslintTester.addRuleTest("lib/rules/one-var", {
                     type: "VariableDeclaration"
                 }
             ]
+        },
+        {
+            code: "function foo() { var bar = true, baz = false; }",
+            args: [2, "never"],
+            errors: [{
+                message: "Split 'var' declaration into multiple statements.",
+                type: "VariableDeclaration"
+            }]
+        },
+        {
+            code: "function foo() { var bar = true; var baz = false; }",
+            args: [2, "always"],
+            errors: [{
+                message: "Combine this with the previous 'var' statement.",
+                type: "VariableDeclaration"
+            }]
+        },
+        {
+            code: "function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }",
+            ecmaFeatures: {
+                destructuring: true
+            },
+            args: [2, "always"],
+            errors: [{
+                message: "Combine this with the previous 'var' statement.",
+                type: "VariableDeclaration"
+            }]
         }
     ]
 });

